### PR TITLE
note viewed and updated date works

### PR DESCRIPTION
**Pull request by Adam3004**

### CODE REVIEW

File: src/main/java/com/brightpath/learnify/persistance/note/NoteRepository.java [hunk 1]
```diff
@@ -1,15 +1,24 @@
...
 @Repository
 public interface NoteRepository extends JpaRepository<NoteEntity, UUID> {
-    List<NoteEntity> findTop4ByOrderByUpdatedAtDesc();
+        @Query("""
+            SELECT u
+            FROM NoteEntity u
+            ORDER BY COALESCE((SELECT ds.viewedAt FROM u.dateStatistics as ds WHERE ds.userId=:userId), u.createdAt)
+            DESC
+            """)
+    List<NoteEntity> findTop4ByOrderByViewedAtDesc(@Param("userId") String userId, Pageable pageable);
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/persistance/note/NoteRepository.java]: Change the name of method to something like `findRecentlyVisitedNotes`

File: src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+@Entity
+@Table(name = "note_date_statistics")
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+public class DateStatisticsEntity {
+    @Id
+    @Column(name = "id", nullable = false, unique = true)
+    private UUID id;
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java:24]: Do we need to assign UUID for those entities? I think we could use standard generated numbers

File: src/main/java/com/brightpath/learnify/persistance/common/PersistentMapper.java [hunk 3]
```diff
@@ -35,20 +37,38 @@ public User asUser(UserEntity entity) {
...
-                entity.getUpdatedAt(),
+                asUpdatedAt(foundStatistics),
+                asViewedAt(foundStatistics),
                 entity.getType(),
                 entity.getPagesCount()
         );
     }
 
+    private OffsetDateTime asUpdatedAt(Optional<DateStatisticsEntity> dateStatistics) {
+        if (dateStatistics.isEmpty()) {
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/persistance/common/PersistentMapper.java]: why not: `foundStatistics.map(DateStatisticsEntity::getUpdatedAt).orElse(null)`?

File: src/main/java/com/brightpath/learnify/domain/note/NoteService.java [hunk 4]
```diff
@@ -61,79 +65,115 @@ public Note createNote(String title, String description, UUID workspaceId, Strin
...
+        List<NoteEntity> notes = noteRepository.findTop4ByOrderByViewedAtDesc(userId, pageable);
         return notes.stream()
-                .map(persistentMapper::asNote)
+                .map(note -> persistentMapper.asNote(note, userId))
                 .filter(note -> permissionAccessService.hasUserAccessToResource(userId, note.id(), NOTE, READ_ONLY))
                 .toList();
     }
 
-    public Note getNoteById(UUID uuid) {
+    public Note getNoteById(UUID uuid, String userId, boolean shouldUpdateViewedDate) {
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/domain/note/NoteService.java]: Can we have two methods here? One for checking if note exist and second for getting note with updating the value?

File: src/main/java/com/brightpath/learnify/domain/binding/BindingService.java [hunk 5]
```diff
@@ -81,8 +81,8 @@ public void removeBindingForNote(UUID noteId) {
         bindingRepository.deleteAllByNoteId(noteId);
     }
 
-    private void checkIfNoteExists(UUID noteId) {
-        noteService.getNoteById(noteId);
+    private void checkIfNoteExists(UUID noteId, String userId) {
```
Comment (krisx303) [inline src/main/java/com/brightpath/learnify/domain/binding/BindingService.java]: Let's add another method in NoteService. It's weird that we have to pass userId to check if entity exists and userId doesn't matter here

**Review by krisx303 — APPROVED** [review]
Generally ok, but I would get rid of UUID from this entity if possible

File: src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java (hunk 2 above)
Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java:24]: or maybe userId?

File: src/main/java/com/brightpath/learnify/domain/note/NoteService.java (hunk 4 above)
Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/domain/note/NoteService.java]: thought about it but chose faster option :p

File: src/main/java/com/brightpath/learnify/persistance/common/PersistentMapper.java (hunk 3 above)
Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/persistance/common/PersistentMapper.java]: nice catch

File: src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java (hunk 2 above)
Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java:24]: ok it cannot be userId beacuse we have it per note, but entity is per everything

Comment (Adam3004) [reply src/main/java/com/brightpath/learnify/persistance/note/date/DateStatisticsEntity.java:24]: i will leave uuid, why numbers?

**Pull request merged**